# STUDENT NAME AND INFORMATION

Note: For the cell above, feel free to use the markdown code from the class Jupyter Notebook and modify it with your information. 

## DATA 601 - Spring 2023
### Homework Assignment 6

#### Due-date: Monday 5/8/2023 midnight.
The excercises in this homework provide instruction on what to do. Feel free to add as many code or markdown cells as you need. The main objective of this homework is to train a classification algorithm and predict a target. In this case we will be using the wine dataset from SKLearn and use various features of wine to predict a target. In this case the target will be defined as the "class". However, note that these algorithms can predict number as well as categorical values. The classification models steps (e.g., train, test, calculate metrics, etc)) discussed in class are the same here.

Wine Dataset:
- https://scikit-learn.org/stable/datasets/toy_dataset.html#wine-recognition-dataset
- https://scikit-learn.org/stable/modules/generated/sklearn.datasets.load_wine.html#sklearn.datasets.load_wine

SciKit Learn Datasets:
- https://scikit-learn.org/stable/datasets

Seaborn Datasets:
- https://seaborn.pydata.org/generated/seaborn.load_dataset.html
- https://github.com/mwaskom/seaborn-data

In [1]:
from sklearn.datasets import load_wine
import pandas as pd
import seaborn as sns

In [5]:
df, target = load_wine(return_X_y=True, as_frame=True) # Loads the data and target
df['wine_class'] = target.tolist() # Adds the target data as a new column in dataframe
df

alcohol  malic_acid   ash  alcalinity_of_ash  magnesium  total_phenols  \
0      14.23        1.71  2.43               15.6      127.0           2.80   
1      13.20        1.78  2.14               11.2      100.0           2.65   
2      13.16        2.36  2.67               18.6      101.0           2.80   
3      14.37        1.95  2.50               16.8      113.0           3.85   
4      13.24        2.59  2.87               21.0      118.0           2.80   
..       ...         ...   ...                ...        ...            ...   
173    13.71        5.65  2.45               20.5       95.0           1.68   
174    13.40        3.91  2.48               23.0      102.0           1.80   
175    13.27        4.28  2.26               20.0      120.0           1.59   
176    13.17        2.59  2.37               20.0      120.0           1.65   
177    14.13        4.10  2.74               24.5       96.0           2.05   

     flavanoids  nonflavanoid_phenols  proanthocyanins  color_intensity   hue  \
0          3.06                  0.28             2.29             5.64  1.04   
1          2.76                  0.26             1.28             4.38  1.05   
2          3.24                  0.30             2.81             5.68  1.03   
3          3.49                  0.24             2.18             7.80  0.86   
4          2.69                  0.39             1.82             4.32  1.04   
..          ...                   ...              ...              ...   ...   
173        0.61                  0.52             1.06             7.70  0.64   
174        0.75                  0.43             1.41             7.30  0.70   
175        0.69                  0.43             1.35            10.20  0.59   
176        0.68                  0.53             1.46             9.30  0.60   
177        0.76                  0.56             1.35             9.20  0.61   

     od280/od315_of_diluted_wines  proline  wine_class  
0                            3.92   1065.0           0  
1                            3.40   1050.0           0  
2                            3.17   1185.0           0  
3                            3.45   1480.0           0  
4                            2.93    735.0           0  
..                            ...      ...         ...  
173                          1.74    740.0           2  
174                          1.56    750.0           2  
175                          1.56    835.0           2  
176                          1.62    840.0           2  
177                          1.60    560.0           2  

[178 rows x 14 columns]

# Additional Python Libraries

In [6]:
# sklearn
# Confusion matrix, Accuracy, sensitivity and specificity
from sklearn.metrics import confusion_matrix 
from sklearn.model_selection import train_test_split

# Classification Models
from sklearn import svm
from sklearn.ensemble import RandomForestClassifier

# Question 1 (5 Points)
Create a new df_filtered dataframe and only include class 0 and 1 in our dataset. We will use this in our classification models below and only predict class 0 or 1.

In [7]:
df_filtered = df[df['wine_class'] < 2].reset_index()
df_filtered

index  alcohol  malic_acid   ash  alcalinity_of_ash  magnesium  \
0        0    14.23        1.71  2.43               15.6      127.0   
1        1    13.20        1.78  2.14               11.2      100.0   
2        2    13.16        2.36  2.67               18.6      101.0   
3        3    14.37        1.95  2.50               16.8      113.0   
4        4    13.24        2.59  2.87               21.0      118.0   
..     ...      ...         ...   ...                ...        ...   
125    125    12.07        2.16  2.17               21.0       85.0   
126    126    12.43        1.53  2.29               21.5       86.0   
127    127    11.79        2.13  2.78               28.5       92.0   
128    128    12.37        1.63  2.30               24.5       88.0   
129    129    12.04        4.30  2.38               22.0       80.0   

     total_phenols  flavanoids  nonflavanoid_phenols  proanthocyanins  \
0             2.80        3.06                  0.28             2.29   
1             2.65        2.76                  0.26             1.28   
2             2.80        3.24                  0.30             2.81   
3             3.85        3.49                  0.24             2.18   
4             2.80        2.69                  0.39             1.82   
..             ...         ...                   ...              ...   
125           2.60        2.65                  0.37             1.35   
126           2.74        3.15                  0.39             1.77   
127           2.13        2.24                  0.58             1.76   
128           2.22        2.45                  0.40             1.90   
129           2.10        1.75                  0.42             1.35   

     color_intensity   hue  od280/od315_of_diluted_wines  proline wine_class  
0               5.64  1.04                          3.92   1065.0          0  
1               4.38  1.05                          3.40   1050.0          0  
2               5.68  1.03                          3.17   1185.0          0  
3               7.80  0.86                          3.45   1480.0          0  
4               4.32  1.04                          2.93    735.0          0  
..               ...   ...                           ...      ...        ...  
125             2.76  0.86                          3.28    378.0          1  
126             3.94  0.69                          2.84    352.0          1  
127             3.00  0.97                          2.44    466.0          1  
128             2.12  0.89                          2.78    342.0          1  
129             2.60  0.79                          2.57    580.0          1  

[130 rows x 15 columns]

# Question 2 (5 Points)
Split the data into training (80%) and testing (20%).

In [8]:
# Defining features to use for prediction (X) and target (y)
x = df_filtered.loc[:,['alcohol','malic_acid','ash','alcalinity_of_ash','magnesium','total_phenols',
              'flavanoids','nonflavanoid_phenols','proanthocyanins','color_intensity','hue',
              'od280/od315_of_diluted_wines','proline']].values
y = df_filtered.loc[:,'wine_class'].values
# Train/Test Split of the data
X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=1)

# Question 3 (10 Points)
Use SVM algorithm to predict the class and calculate metrics: Confustion Matrix, Accuracy, Sensitivity and Specificity.  

In [9]:
clf = svm.SVC()
clf.fit(X_train, y_train)

y_predict = clf.predict(X_test)

cm1 = confusion_matrix(y_test, y_predict)
print('Confusion Matrix : \n', cm1)
total1=sum(sum(cm1))
#neigh.score(X_test, y_test) # Accuracy from the .score function.
#####from confusion matrix calculate accuracy
accuracy1 = (cm1[0,0] + cm1[1,1]) / total1
print ('Accuracy : ', accuracy1)

sensitivity1 = cm1[0,0] / (cm1[0,0] + cm1[0,1])
print('Sensitivity : ', sensitivity1)

specificity1 = cm1[1,1] / (cm1[1,0] + cm1[1,1])
print('Specificity : ', specificity1)

Confusion Matrix : 
 [[12  1]
 [ 0 13]]
Accuracy :  0.9615384615384616
Sensitivity :  0.9230769230769231
Specificity :  1.0


# Question 4 (10 Points)
Use a random forest classifier to predict the class and calculate metrics: Confustion Matrix, Accuracy, Sensitivity and Specificity.

In [10]:
clf = RandomForestClassifier(max_depth=2, random_state=0)
clf.fit(X_train, y_train)

y_predict = clf.predict(X_test)

cm1 = confusion_matrix(y_test, y_predict)
print('Confusion Matrix : \n', cm1)
total1=sum(sum(cm1))
#neigh.score(X_test, y_test) # Accuracy from the .score function.
#####from confusion matrix calculate accuracy
accuracy1 = (cm1[0,0] + cm1[1,1]) / total1
print ('Accuracy : ', accuracy1)

sensitivity1 = cm1[0,0] / (cm1[0,0] + cm1[0,1])
print('Sensitivity : ', sensitivity1)

specificity1 = cm1[1,1] / (cm1[1,0] + cm1[1,1])
print('Specificity : ', specificity1)

Confusion Matrix : 
 [[13  0]
 [ 0 13]]
Accuracy :  1.0
Sensitivity :  1.0
Specificity :  1.0


# Question 5 (5 Points)
How many datapoints do we have in this dataset? Provide thoughts and comments on the size of this dataset on at least two of the following:  
- Which of the classification models above performed better?
- What are some challenges of the size of this dataset?
- Would you feel comfortable with using this dataset for predictions?
- What could be some things that you could do for improving the outputs and performance of the models? This could include steps during the data cleaning/preparation/transformation, testing or training steps.

In [11]:
print(f'The wine dataset has {df.shape[0]} rows of data')

The wine dataset has 178 rows of data


In [12]:
df_filtered['wine_class'].value_counts()

1    71
0    59
Name: wine_class, dtype: int64

Full Answer:

Of the two models SVM and Random Forests, Random Forests seems to have performed better.

This dataset seems to be on the smaller side. It may make sense to evaluate with a subject matter expert or perform some other evaluations to make sure the size of this dataset is appropriate for a classification task. There may be a need to increase the amount of data and depending on how easy it is to obtain further data, it may make sense to do so.

The data has three classes and only 178 records total. The value counts of the wine class 0 above shows that at best we have 71 wines with class 0. This means that during the train/test split of the data for this class only about 14 records would be part of the test data while the rest training. This is a really small number of records to perform testing, and calculate metrics and may not be a statistically significant sample.

If we would have performed cleaning/preparation/transformation of the data one task that could have improved the outputs and the performance, these include:
- evaluate need for data cleaning (e.g., removing outliers, missing values, etc.)
- scaling of the data
- feature selection

Because the amount of data is so low we may not want to drop any rows of data. For scaling, there are few features (e.g., proline) that have relatively large values compared to the others. Feature selection may also improve the model predictive performance. Because the accuracy is so high and the amount of data so low, I don't expect that either scaling or feature selection will make any difference.

We could also expand to do multi class or multi label classification instead of separating and do the binary classification.

# Classification with SCALING AND MULTILABEL

In this section we are exploring how to use scaling and multilabel classification including the multilabel classification matrix. SKLearn scaling functions and methods include MinMaxScaler, StandardScaler, MaxAbsScaler, RobustScaler

SKLearn Scalers:
- https://scikit-learn.org/stable/auto_examples/preprocessing/plot_all_scaling.html

Multilabel Classification Confusion Matrix
- https://scikit-learn.org/stable/modules/generated/sklearn.metrics.multilabel_confusion_matrix.html

In [13]:
from sklearn.preprocessing import MinMaxScaler # Other SKLearn scalers StandardScaler, MaxAbsScaler, RobustScaler
from sklearn.metrics import multilabel_confusion_matrix

In [18]:
# Defining features to use for prediction (X) and target (y)
scaler = MinMaxScaler() #Creating StandardScaler Object
# Note that scaling can also be performed during fitting step.
df_scaled = df.copy() # Creating a copy of the dataframe
df_scaled.iloc[:,:-1] = scaler.fit_transform(df_scaled.iloc[:,:-1]) # Scaling everything except target column

# The Target needs to be an Object or will give an error in the multilabel confusion matrix.
df_scaled['wine_class'] = df_scaled['wine_class'].astype(str)
df_scaled # Scaled data.

alcohol  malic_acid       ash  alcalinity_of_ash  magnesium  \
0    0.842105    0.191700  0.572193           0.257732   0.619565   
1    0.571053    0.205534  0.417112           0.030928   0.326087   
2    0.560526    0.320158  0.700535           0.412371   0.336957   
3    0.878947    0.239130  0.609626           0.319588   0.467391   
4    0.581579    0.365613  0.807487           0.536082   0.521739   
..        ...         ...       ...                ...        ...   
173  0.705263    0.970356  0.582888           0.510309   0.271739   
174  0.623684    0.626482  0.598930           0.639175   0.347826   
175  0.589474    0.699605  0.481283           0.484536   0.543478   
176  0.563158    0.365613  0.540107           0.484536   0.543478   
177  0.815789    0.664032  0.737968           0.716495   0.282609   

     total_phenols  flavanoids  nonflavanoid_phenols  proanthocyanins  \
0         0.627586    0.573840              0.283019         0.593060   
1         0.575862    0.510549              0.245283         0.274448   
2         0.627586    0.611814              0.320755         0.757098   
3         0.989655    0.664557              0.207547         0.558360   
4         0.627586    0.495781              0.490566         0.444795   
..             ...         ...                   ...              ...   
173       0.241379    0.056962              0.735849         0.205047   
174       0.282759    0.086498              0.566038         0.315457   
175       0.210345    0.073840              0.566038         0.296530   
176       0.231034    0.071730              0.754717         0.331230   
177       0.368966    0.088608              0.811321         0.296530   

     color_intensity       hue  od280/od315_of_diluted_wines   proline  \
0           0.372014  0.455285                      0.970696  0.561341   
1           0.264505  0.463415                      0.780220  0.550642   
2           0.375427  0.447154                      0.695971  0.646933   
3           0.556314  0.308943                      0.798535  0.857347   
4           0.259386  0.455285                      0.608059  0.325963   
..               ...       ...                           ...       ...   
173         0.547782  0.130081                      0.172161  0.329529   
174         0.513652  0.178862                      0.106227  0.336662   
175         0.761092  0.089431                      0.106227  0.397290   
176         0.684300  0.097561                      0.128205  0.400856   
177         0.675768  0.105691                      0.120879  0.201141   

    wine_class  
0            0  
1            0  
2            0  
3            0  
4            0  
..         ...  
173          2  
174          2  
175          2  
176          2  
177          2  

[178 rows x 14 columns]

In [19]:
# Defining features to use for prediction (X) and target (y)
# Note we are using the full dataset and in this case using the index for feature selection and not using the values.
x = df_scaled.iloc[:,:-1]
y = df_scaled.iloc[:,13]
# Split of the data
X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=1)

Note that if we are reussing a code block like in this case we are reussing the models and accuracy metrics. It makes sense to define a custom function rather than retyping like I am doing in this case.

In [23]:
clf = svm.SVC()
clf.fit(X_train, y_train)

y_predict = clf.predict(X_test)

cm1 = confusion_matrix(y_test, y_predict)
print('Confusion Matrix : \n', cm1)
total1=sum(sum(cm1))
#neigh.score(X_test, y_test) # Accuracy from the .score function.
#####from confusion matrix calculate accuracy
accuracy1 = (cm1[0,0] + cm1[1,1]) / total1
print ('Accuracy : ', accuracy1)

sensitivity1 = cm1[0,0] / (cm1[0,0] + cm1[0,1])
print('Sensitivity : ', sensitivity1)

specificity1 = cm1[1,1] / (cm1[1,0] + cm1[1,1])
print('Specificity : ', specificity1)

# MultiLabel Confusion Matrix:
multilabel_confusion_matrix(y_test, y_predict)
# Note that either of the classification matrix are correct. It is different ways to visualize the data.

Confusion Matrix : 
 [[14  0  0]
 [ 0 13  0]
 [ 0  1  8]]
Accuracy :  0.75
Sensitivity :  1.0
Specificity :  1.0


array([[[22,  0],
        [ 0, 14]],

       [[22,  1],
        [ 0, 13]],

       [[27,  0],
        [ 1,  8]]], dtype=int64)

In [24]:
clf = RandomForestClassifier(max_depth=2, random_state=0)
clf.fit(X_train, y_train)

y_predict = clf.predict(X_test)

cm1 = confusion_matrix(y_test, y_predict)
print('Confusion Matrix : \n', cm1)
total1=sum(sum(cm1))
#neigh.score(X_test, y_test) # Accuracy from the .score function.
#####from confusion matrix calculate accuracy
accuracy1 = (cm1[0,0] + cm1[1,1]) / total1
print ('Accuracy : ', accuracy1)

sensitivity1 = cm1[0,0] / (cm1[0,0] + cm1[0,1])
print('Sensitivity : ', sensitivity1)

specificity1 = cm1[1,1] / (cm1[1,0] + cm1[1,1])
print('Specificity : ', specificity1)

# MultiLabel Confusion Matrix:
multilabel_confusion_matrix(y_test, y_predict)
# Note that either of the classification matrix are correct. It is different ways to visualize the data.

Confusion Matrix : 
 [[14  0  0]
 [ 1 12  0]
 [ 0  0  9]]
Accuracy :  0.7222222222222222
Sensitivity :  1.0
Specificity :  0.9230769230769231


array([[[21,  1],
        [ 0, 14]],

       [[23,  0],
        [ 1, 12]],

       [[27,  0],
        [ 0,  9]]], dtype=int64)

# BONUS (5 Points)

We have the dataset below related to diamonds. 
1. Which feature could we predict?
2. Which model would be an appropriate model to use?

In [ ]:
diamonds = sns.load_dataset("diamonds")
diamonds.head()

We could predict the price of the diamonds given the other features. We would need to use a model that can be used to predict continuous numeric values. The simplest model that we could use is a multiple linear regression.